juju/juju
# Application API Client Support for Config with --generation Option

## Description of change

This patch threads the active generation through API clients from application commands.

- Commands for _deploy_ (including bundles), _bundle-diff_ and _set-plan_ always assume the current generation.
- Commands for application and charm config allow the user to add a _--generation_ option, which if not set will use the active generation from from local storage (defaulting to "current" here).
- Resources are not dealt with by the patch but will be generational in future.

Patches will follow this one to:
- Retrieve the active generation from local storage (depends on #9661).
- Ensure all usages of generation are passed in calls to the API server.

## QA steps

At this time, we want to ensure that all client operations involving generational settings are operational, but not backward-breaking. So we test a new client against an old controller.

- Bootstrap a controller with a version prior to this patch - develop HEAD will do fine.
- Build and install Juju from this patch.
- `export JUJU_DEV_FEATURE_FLAGS=generations`
- Deploy an old version of an application, for example `juju deploy cs:trusty/mariadb-6`
- Check that the following commands work correctly with and without the _--generation_ option:
  - `juju config mariadb` (plus get, set and unset).
  - `juju upgrade-charm` (should cause an upgrade in this example to cs:trusty/mariadb-7).
  - `juju export-bundle > temp.yaml` runs without error.
  - `juju diff-bundle temp.yaml` runs without error.
  - `juju set-plan mariadb www.google.com` will return an error due to the URL, but will run correctly.

## Documentation changes

To come with _Generations_ feature generally.

## Bug reference

N/A
